### DISCUSSION TITLE
Potential Types Updates

### DISCUSSION THREAD

**Discussion by icylace** [jorgebucaran/hyperapp#1052] [Ideas]
Hyperapp's TypeScript definitions are more or less solid at this point. However, a few topics remain that are a little more about philosophical positions rather than technical advancement.

Here they are in the priority I think they should be in.

1. Loosen the return type for effecters?
2. Is it worth adding a `TypedText` for `text()` as there is a `TypedH` for `h`?
3. Rename `app()`'s `dispatch:` property? (not just a types question)
4. Potentially inline more types the user isn't expected to use directly. Specifically: `DispatchInitializer`, `Key`, `Subscriptions`, `Tag`, `Unsubscribe`, and `VDOMNodeType`.
5. Is `View` a worthwhile type definition?
6. Must/should `Unsubscribe` be part of `Subscriptions`?

Here is some other stuff that is also important but probably doesn't need as much debate.

- Make more code examples.
- How many of the `any`s currently being used can be replaced by more type-safe `unknown`s while preserving DX?
- Complete the documentation for Hyperapp's TypeScript support.

Update (4/20/2021, 8:05 PM EST):

7. What types should be made private?

**Comment by icylace** [comment]
Here are my thoughts for the numbered ones:

1. I prefer the return type to remain `void | Promise<void>`. This lets effecters be `async` functions, which I find convenient, while emphasizing that effecters do indeed cause side-effects in general because `void`-returning functions are almost always like that. However, since Hyperapp takes a conceptually hands-off approach to effecters I can accept `unknown` or in the worst case `any`.

2. Maybe `TypedText` is worth it but not sure. I'm wondering if there's a situation where setting the state type parameter of `h()` or using a `TypedH` version of `h()` wouldn't be enough for TypeScript's inference to determine the state type for `text()`. Even if there isn't there's the DX to consider.

3. To leave it named as `dispatch:` would invite confusion for newcomers. I understand that going back to `middleware:` is not ideal. The name I'm currently thinking of is `initDispatch:`. Boring, yes, but the name expresses this function's role of initiating the dispatch function.

4. If the goal is to keep the initial number of types as small as possible based upon estimated chance of usage, then yes inline them all with the exception of `Unsubscribe`. That one would get used in the return type of users' subscriber functions. Even though `Unsubscribe` is easy for the user to make, it feels more like code-level documentation rather than a convenience type. Inlining it to `() => void` or `() => undefined` wouldn't help code comprehension.

5. Maybe not. Keeping it as-is (`type View<S> = (state: S) => VDOM<S>`) would preclude usage of [returning multiple VNodes](https://github.com/jorgebucaran/hyperapp/blob/main/docs/architecture/views.md#components-returning-multiple-vnodes). On the other hand, using `type View<S> = (state: S) => VDOM<S> | VDOM<S>[]` would complicate nested views. Introducing additional view-related types instead would also complicate things. But `const foo: View<State> = (state) => ...` is not much smaller than the equivalent `const foo = (state: State): VDOM<State> => ...`. 

    In defense of `View`, it's probably at its best when used in the parameter list of view wrapping functions that act as decorators, or more involved higher-order components. Also, `(state: S) => VDOM<S> | VDOM<S>[]` and `(state: S) => VDOM<S>[]` can still be made regardless of `View`. They would just only be technically "view-like".

6. If it turns out that `Unsubscribe` can technically be left out of `Subscriptions` I would be okay leaving it out since it is an implementation detail. Also, I can't think of a scenario where directly putting an unsubscriber function in the subscriptions array makes sense from the user's perspective.

**Reply by zaceno** [reply]
I don't have super strong opinions on these, but here are a few of my thoughts.

1. I know people generally try to avoid `any` because it often means "I give up". But in this case it is actually the correct type. An effecter could return the king james bible for all hyperapp cares.  On the other hand, there is no value to returning something from an effecter and it's kind of weird if you do. A "code smell". So from that perspective `void | Promise<void>` makes sense. 

I guess the broader question is how much we want the types to offer guidance beyond correctness, or just be correct. I lean toward not offering guidance and just being correct, in general. Except when there is no value in doing something that is possible, then doing it anyway is probably a mistake and it makes sense to have the types help you here. 

So: is there ever any reason anyone might want to return something from an effecter? Hyperapp doesn't care of course, but maybe there is some other reason? If there is, I say we go with `any` .If there is not, I say we go with `void | Promise<void>`

**Reply by zaceno** [reply]
2. I still haven't quite understood TypedH or when I would need it. So right now I'm leaning toward taking out TypedH, rather than adding a TypedText.

**Reply by zaceno** [reply]
4. I need to experiment more with implementing typed subscirptions to say anything about unsubscribe. But generally I am somewhat averse to the idea of types whose sole purpose is for self-documenting code et c.

It's not that I'm against self-documenting code, but I want to be in control of names.  Of course there is nothing preventing me from inventing my own names for types that are already defined (one of the great boons of structural typing) – but it's what makes me see little value in those predefined types.

**Reply by zaceno** [reply]
5. If we were to change `View` to return `MaybeVDOM | MaybeVDOM[]` it could no longer be used as a type in the definition of the `App` type. Becuase `App.view` _must_ return exactly `VDOM` and nothing else. 

And if we inline the definition of `App.view` and offer `View` as a utility type for views then I think it will be not very useful. It would be much more useful to have a type for reusable view components, something like:

```ts
type ViewComponent<S, Props> = <X>(props: ValidateCustomPayloads<S, X> & Props, content: MaybeVDOM<S>[]) => MaybeVDOM<S>
```

but even that is not the job of the core type definitions to provide imo. So I am _not_ suggesting we add that. Rather I'm on the side of inlining and not exposing the `View` type (if the results are reasonably readable).

**Reply by zaceno** [reply]
6. Agree

**Reply by icylace** [reply]
> 2. I still haven't quite understood TypedH or when I would need it. So right now I'm leaning toward taking out TypedH, rather than adding a TypedText.

The full discussion with relevant code examples is [here](https://github.com/jorgebucaran/hyperapp/issues/1048) but I'll try to boil it down.

Imagine you're working on a large and potentially complicated set of view functions that are arranged in some sort of hierarchy. `h()` doesn't know and can't know ahead of time about your application state, so TypeScript's inference engine is relied upon to give your `h()` calls the proper state type.

However, TypeScript can't do that in all cases which is why explicit assignment (e.g. `h<State>("p", {}, ...)`) is sometimes necessary. The good news is you only need to do that for certain `h()` calls and the inference engine can handle the rest. However, figuring out which `h()` calls to do this for sometimes feels like it's more effort than it's worth.

If that wasn't annoying enough, consider the effect of refactoring over time. The `h()` calls that have the explicit state type get moved to different parts of the view hierarchy and are no longer in a position to be helpful to the inference engine and their explicit state type assignments have become redundant. Simultaneously, different `h()` calls need explicit state type assignments. Managing this is at best annoying and at worst confusing. Either way you're faced with some friction that slows you down due to how the language works.

`TypedH` solves this by letting you create an app-specific version of `h()` that is fully aware of your app's state type. This saves you from having to worry about any of this stuff if you don't think it's worth it.

**Reply by zaceno** [reply]
Thanks for the explanation, I understand now. I agree that it is not always easy to know when you need to add the state specifier. I can see how this would be useful if you're making views with plain h/text. And it's complex enough it's worth offering rather than expecting people to write themselves. 

So I'm ok leaving it in for the time being. But really I'd rather try to work out exactly how and when the inference fails and find some kind of general solution for it. Because the same problem occurs with my typed version of @hyperapp/html and I am also starting to look in to how to use the types with hyperlit & JSX (it's not looking good 😢 ...). But that's a whole other topic.  

About TypedText I think if you are already using TypedH, I'm pretty sure you will never also need TypedText. So I think we're good.

**Reply by icylace** [reply]
Cool, I'm good with not having a `TypedText`.

I don't know exactly how TypeScript's inference engine works but I do know that having a sound type system [isn't a priority](https://www.typescriptlang.org/docs/handbook/type-compatibility.html#a-note-on-soundness) for the language.

For better or worse we're not talking about something on the level of Haskell here 😅

**Reply by icylace** [reply]
Re:  number 4, the benefit of having predefined **canonical** names is for communication with other people. The better the choice of what gets named and what the name is, the better the things like the following are:

- Discussing details with fellow team members during code reviews.
- Explaining how to use Hyperapp + TypeScript in blog posts.
- Posting Hyperapp-related questions on StackOverflow.
- etc.

**Reply by icylace** [reply]
I think it becomes something to consider when the type in question would get used by users, as is the case for `Unsubscribe`.

**Reply by icylace** [reply]
Re: number 5, I agree with inlining `View`. Letting go of it seems to let us sidestep a lot of awkwardness.

Also, that `ViewComponent` type looks pretty useful.  I'll be borrowing it 😄

**Comment by zaceno** [comment]
I'm back to thinking about `TypedH`.

 I was updating my [types usage examples](https://gist.github.com/zaceno/e25fe832a1e11168b05dbce4a16b832b) gist to show how to deal with inference troubles in view components. I first use the simple example where explitly declaring the return type as `VDOM<S>` solves the problem. Then I was going to use TypedH on a different example, but I couldn't come up with one. 🤔  
 
 I mean, TypedH could also have solved the problem, but in a less easy way. Do we have some example where just declaring the return type isn't enough?

For easy reference, here is the example I used:

```ts
import {h, text, Action, VDOM} from "hyperapp"

type MyComponentProps<S> = {
  foo: Action<S, any>, 
  bar: string
}

// In this version, typescript tries to infer the
// return type of the view component, but cannot,
// because only the first child has any actions 
// from which the state type can be inferred
// in the second child, the state type is inferred
// to be unknown.
const myComponent = <S>(props: MyComponentProps<S>) =>
  h('div', {}, [
    h('button', {onclick: props.foo}, text('click')),
    h('p', {}, text(props.bar)),
  ])

// The easiest way to solve this is to explicitly
// declare the return type:
const myComponent2 = <S>(props: MyComponentProps<S>):VDOM<S> => 
  h('div', {}, [
    h('button', {onclick: props.foo}, text('click')),
    h('p', {}, text(props.bar)),
  ])

```

[playground](https://www.typescriptlang.org/play?jsx=0&ssl=1&ssc=1&pln=29&pc=1#code/JYWwDg9gTgLgBAbwBYBo4wKYA8ZoIIDGMwEAdmgGoAiA8gLIC+cAZlBCHAERICeYGUAIZgwnAFBiYfDHDo8Awu0ikMpGAAU2YAM4AeAMoA+OAF5EYuCwgQAXHELEyBtINI9DaC3ABGgqHe0YKGBSAHMxBgkAeii4AElSdCRgbTgANwFtEnJ0aW0CYLB4IOAMVJgIOBDmASSMMRi4KAwYAFcoRKl+OAhmOvTSgHc4AiUyVVwfVvgCV1IIXAbY7wxZ1u0ZMgAbHn7mYChAkeStgBM4JEFU113BImzUpZY2DkHkgiR+wMFMXO7ZxIrKqkGpQZqnJ4hL6rMjnD7AM5oGBIGTfX5dGQpYGg8FPCo+GStUgAa3mg1IADoxKNSEcQAoxio1KY4AZDAAKMBabR2OSKcDjNSaCA6NkASlMhi8SHZAHJTsA0rK0AgGGgANpeSwy2XeaYVUjKxBkAhbYAEYl2Lki7QU5jWNXobAwOWm83E2VisUoLUXOVgI2qpHOzncim+KBen2WAC6YuisQAKii4BgrqUjoNBLt8doIFsMkksVj8dgwGaCMAYDsnqdVls-DJkTJmm0On8MDZqWQ6QyBUyYAAmFls0M23l95QTYWioximzUehsyVwaVyhVKlWOzWWbVyvUwA2Bk0Vy1wa06O0O4M4V2nz3e306gNbm8ui+2iNRrxxiRiIA)

EDIT: The issue in #1048 doesn't seem to be solved quite so easily. Investigating further

**Reply by zaceno** [reply]
I haven't yet got this to work in the main view of my apps (so there I still need to sprinkle in `<State>` here and there. But I wonder if there isn't a way to use this approach there too. 🤔

**Reply by icylace** [reply]
I think it's great you found another solution but it's not necessarily easier. What you mention in https://github.com/jorgebucaran/hyperapp/issues/1048#issuecomment-822858741 as an alternative solution is to use a combination of explicitly stating the `VDOM` returns and using `map` in a certain way. Always stating the `VDOM` returns, while arguably good practice for different reasons, comes at the cost of verbosity and scales poorly compared to the `TypedH` solution. And that `map` usage is especially subtle and would otherwise be considered verbose as well.

I didn't know about the `map` thing, so thank you for mentioning it but that just reinforces my point about having to worry about such things. Perhaps `reduce`, `filter`, etc. are affected similarly? Maybe other types of methods need to be treated in a special way too?

In contrast, 
```ts
import { TypedH, h as ha } from "hyperapp"

type State = { youGetTheIdea: boolean }

const h: TypedH<State> = ha
```
and now I'm **done** worrying about it without making further compromises.

**Reply by zaceno** [reply]
Sure I see. It's a neat way to "brute-force" inference issues typescript might get caught up on but shouldn't. I was mainly interested in an example where TypedH is the only "right way" to solve the problem. Now I understand TypedH is not about being the "right way" but a fallback when the right way is elusive or not worth the time. I could probably use an example with `map` to demonstrate the kind of pitfalls typescript throws at you, and that TypedH lets you sidestep.

I'm personally more interested in using other view-methods like for instance html-functions.  As far as I can tell TypedH can't help me there. (I'd need to define a TypedA, TypedB, TypedEm, TypedSpan, TypedP.... ) So I'm still pretty interested in understanding more how and when the inference can fail, and how to solve it. It can't be completely random after all.

**Reply by icylace** [reply]
That's a good point about view methods. My solution to that, which I've just implemented in my `ntml` package, is to include an alternative version of my element factory function that requires a state type.

So, compare [`n`](https://github.com/icylace/ntml/blob/f9133841e28f4f38c82b27226bc310acc8385de0/src/index.ts#L20) versus [`typedN`](https://github.com/icylace/ntml/blob/f9133841e28f4f38c82b27226bc310acc8385de0/src/index.ts#L31). Also check out my rudimentary [test file](https://github.com/icylace/ntml/blob/master/test/index.test-d.ts) which tries out both.

I'm sure you could do something similar for your own library.

**Reply by zaceno** [reply]
that's an interesting idea to have a typed factory. It means you have to create all the tag-functions using the typedN, but I'm sure it wouldn't be too hard to make something more expedient like:

```ts
const {div, span, p, a} = typedNMulti('div', 'span', 'p', 'a')
```

**Reply by zaceno** [reply]
Still personally more interested in working out the when and why inference fails. But I get that this technique is very useful for those who just want to move on with their lives and not worry about it 😅

**Comment by icylace** [comment]
I recently discovered that types _can be private_. I feel silly for not realizing this sooner !

[Like I said](https://discord.com/channels/804672552348680192/805746611467583499/834216698834845706) in Discord, we can just move the types we want private outside of the `declare module "hyperapp"` block and also don't explicitly export them.

Now we don't need to feel pressured into inlining just for the sake of minimizing public exposure for the types. So, the question becomes, what to make private?  Perhaps what would otherwise get inlined should instead be made private.

Also, should we reintroduce types that were inlined but make them private? Maybe their inclusion makes the the types file better structured.

At the moment I think `VDOMNodeType` and `EventsMap` are good as private types. I'll consider the rest as well.

**Reply by jorgebucaran** [reply]
> Also, should we reintroduce types that were inlined but make them private?

Once the types are done and sound, maybe, but not right now, as things are still in flux.

**Reply by icylace** [reply]
I've successfully experimented with making these types private and think they're better that way:

|                       |                |             |
| --------------------- | -------------- | ----------- |
| `AllOrNothing`        | `AtLeastOne`   | `ClassProp`
| `DispatchInitializer` | `EventsMap`    | `IndexableByKey`
| `Key`                 | `StyleProp`    | `Subscriptions`
| `Tag`                 | `VDOMNodeType` | `View`

Which means I reduced the set of public types to:

|                          |                |             |
| ------------------------ | -------------- | ----------- |
| `app()`                  | `h()`          | `memo()`
| `text()`                 | `Action`       | `App`
| `Dispatch`               | `Dispatchable` | `Effect`
| `EventActions`           | `MaybeVDOM`    | `PropList`
| `Subscription`           | `TypedH`       | `Unsubscribe`
| `ValidateCustomPayloads` | `VDOM`

Of those, I'm questioning `PropList` and `ValidateCustomPayloads`. It's not that they aren't essential, it's whether or not they're valuable independent of each other. Perhaps it's worth trying to unify them ?

**Reply by icylace** [reply]
The more I think about it, I see `App` as a private type, too.

**Reply by icylace** [reply]
I'll probably still inline some types because I'm not 100% sure the private types thing is reliable.

**Reply by icylace** [reply]
Oh, idea!  I'm gonna experiment with keeping `App` public for the primary reason of having stuff like `App<State>["view"]` available to the user.  So, `DispatchInitializer` becomes `App<State>["dispatch"]`, etc.

**Reply by zaceno** [reply]
Agree with all of these  except `ClassProp` & `StyleProp`. I think they ought to be public – I sometimes do components that "enhance" other components by adding classes / styles on top of already given class/style props. So I think those types would be useful in those cases (and kind of tricky to define oneself).

**Reply by zaceno** [reply]
Oh and are we really sure `Unsubscribe` needs to be public? I haven't played much with typing subs yet, but seems that might be one to make private.

**Reply by icylace** [reply]
Well-formed subs return an unsubscribe function. Seems like a clear reason to have `Unsubscribe` public.

**Reply by zaceno** [reply]
I suppose. It isn't technically necessary from what I've found (I may have missed some situations) but if a person wanted to explicitly declare their subscriber's return value, then I can see how using `Unsubscribe` reads better than `() => any`

**Reply by icylace** [reply]
Yes, exactly. Every other inlining I've done had some sort of descriptive part to it. For example, inlining `ActionWithPayload` retained descriptiveness because of named tuples. I've also been experimenting with inlining `View` and that seems to be okay because its function signature is a decent hint at what it does. And in most cases it's associated with a `view:` property to further describe it.

**Reply by icylace** [reply]
So, in my further experimentations private types are working consistently.

The reason why I thought they were unreliable was because I was using an ugly hack at one point. I was using `import type { ... } from "hyperapp"` within the `.d.ts` file itself to ensure internal access to some types and that worked but made all the types public. Since then I arranged the types so that that imperfect hack was no longer necessary. I removed the hack and the private types became once again non-importable in user code.
